### PR TITLE
Improve backend test coverage, add CI workflow, and introduce testable design with mocking

### PR DESCRIPTION
**Pull request by TwilightVoyager777**
# Description of Changes

## What was changed

This pull request improves the testing infrastructure and testability of the Stirling-PDF backend.

Key changes include:

* Added multiple **unit tests and white-box tests** for backend components such as:

  * RotationController
  * SecretMasker
  * GlobalExceptionHandler
  * MobileScannerController
  * LanguageService
* Extended existing tests to cover **edge cases and boundary conditions** (e.g., rotation angles and validation logic).
* Implemented **FSM-based testing** to improve coverage for certain features.
* Introduced **mocking and stubbing techniques** to isolate dependencies and improve test reliability.
* Added helper methods and small refactors to support **testable design**.
* Added a **GitHub Actions CI workflow** to automatically build the backend and run core tests.
* Cleaned up CI configuration by removing unused workflows.
* Minor fixes and improvements including Gradle updates and small bug fixes discovered during testing.

## Why the change was made

These changes were made to improve the **test coverage, reliability, and maintainability** of the backend system.

The work applies several software testing practices including:

* White-box testing
* FSM-based testing
* Mocking and stubbing
* Continuous Integration (CI)

By adding automated tests and CI validation, the project can detect regressions earlier and ensure that important backend components behave correctly.

## Challenges encountered

* Some components were originally **tightly coupled**, which made them difficult to test directly.
* Mocking certain dependencies required careful configuration to avoid unintended side effects.
* Configuring the CI workflow to run only the **necessary backend tests** required adjustments to the workflow setup.

---

### General

* [x] I have read the [[Contribution Guidelines](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/CONTRIBUTING.md)](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/CONTRIBUTING.md)
* [x] I have read the [[Stirling-PDF Developer Guide](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/DeveloperGuide.md)](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/DeveloperGuide.md) (if applicable)
* [ ] I have read the [[How to add new languages to Stirling-PDF](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/HowToAddNewLanguage.md)](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/HowToAddNewLanguage.md) (if applicable)
* [x] I have performed a self-review of my own code
* [x] My changes generate no new warnings

### Documentation

* [ ] I have updated relevant docs on [[Stirling-PDF's doc repo](https://github.com/Stirling-Tools/Stirling-Tools.github.io/blob/main/docs/)](https://github.com/Stirling-Tools/Stirling-Tools.github.io/blob/main/docs/) (if functionality has heavily changed)
* [ ] I have read the section [[Add New Translation Tags](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/HowToAddNewLanguage.md#add-new-translation-tags)](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/HowToAddNewLanguage.md#add-new-translation-tags) (for new translation tags only)

### Translations (if applicable)

* [ ] I ran [`[scripts/counter_translation.py](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/docs/counter_translation.md)`](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/docs/counter_translation.md)

### UI Changes (if applicable)

* [ ] Screenshots or videos demonstrating the UI changes are attached (e.g., as comments or direct attachments in the PR)

### Testing (if applicable)

* [x] I have tested my changes locally. Refer to the [[Testing Guide](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/DeveloperGuide.md#6-testing)](https://github.com/Stirling-Tools/Stirling-PDF/blob/main/devGuide/DeveloperGuide.md#6-testing) for more details.

### CODE REVIEW

**Comment by Frooodle** [comment]
Appreciate all the testing however you have made several changes to our GHAs which should not be done, please revert all GHA changes

Thanks again for the added test cases though! Certainly something we need more of

**Comment by Frooodle** [comment]
Once these have been restored will review more

**Comment by TwilightVoyager777** [comment]
sure I will do that soon

File: .github/workflows/build.yml
Comment (balazs-szucs) [inline .github/workflows/build.yml:79]: So, this has caused many headaches in the past, I believe we prefer not to cache via gradle. Frooodle is working some docker caching parrallel to this I believe

Comment (balazs-szucs) [inline .github/workflows/build.yml:164]: SAme comment as above

File: .github/workflows/multiOSReleases.yml
Comment (balazs-szucs) [inline .github/workflows/multiOSReleases.yml:25]: Unnecesary (?)

File: build.gradle [hunk 1]
```diff
@@ -2,11 +2,12 @@ plugins {
     id "java"
     id "jacoco"
     id "io.spring.dependency-management" version "1.1.7"
-    id "org.springframework.boot" version "4.0.3"
+    id "org.springframework.boot" version "3.5.7"
     id "org.springdoc.openapi-gradle-plugin" version "1.9.0"
     id "io.swagger.swaggerhub" version "1.3.2"
     id "com.diffplug.spotless" version "8.1.0"
     id "com.github.jk1.dependency-license-report" version "3.0.1"
+    id "com.github.spotbugs" version "6.1.7" apply false
```
Comment (balazs-szucs) [inline build.gradle:10]: Non-existent version according to: https://mvnrepository.com/artifact/com.github.spotbugs/spotbugs
https://github.com/spotbugs/spotbugs/releases

File: build.gradle [hunk 2]
```diff
@@ -346,6 +319,19 @@ subprojects {
...
+        reportLevel = "low"
+    }
+
+    tasks.withType(com.github.spotbugs.snom.SpotBugsTask).configureEach {
+        reports {
+            xml.required.set(true)
+            html.required.set(true)
+        }
+    }
+
```
Comment (balazs-szucs) [inline build.gradle:334]: Can you test this and add some hints, how this supposed to work?

File: app/core/src/test/java/stirling/software/SPDF/service/LanguageServiceBasicTest.java [hunk 3]
```diff
@@ -103,10 +104,48 @@ void testGetSupportedLanguages_WithRestrictions() {
...
+                };
+
+        when(applicationProperties.getUi().getLanguages()).thenReturn(Collections.emptyList());
+
+        Set<String> supportedLanguages = failingService.getSupportedLanguages();
+
+        assertTrue(supportedLanguages.isEmpty(), "On IO failure, service should return empty set");
+    }
+
+    // Added by Pengcheng Xu: exercise the new resolver seam to avoid subclass-based stubbing.
```
Comment (balazs-szucs) [inline app/core/src/test/java/stirling/software/SPDF/service/LanguageServiceBasicTest.java:131]: We don't really do `@author` javadoc or similar, perhaps these can omitted. (there is lot of author comments)

File: app/core/src/test/java/stirling/software/common/controller/JobControllerTest.java [hunk 4]
```diff
@@ -409,31 +407,235 @@ void testCancelJob_Unauthorized() {
...
 
     @Test
-    void testGetFileMetadata_ForbiddenWhenFileOwnedByAnotherUser() throws Exception {
-        String fileId = "file-id";
+    void testGetJobResult_CompletedSuccessWithMultipleFiles_ReturnsMetadata() {
+        String jobId = "job-with-multiple-files";
+        JobResult mockResult = new JobResult();
+        mockResult.setJobId(jobId);
+        mockResult.completeWithFiles(
+                java.util.List.of(
```
Comment (balazs-szucs) [inline app/core/src/test/java/stirling/software/common/controller/JobControllerTest.java:432]: `java.util.List.of`

FQN. I'm planning to make a PR for this, generally FQN is considered bad style

File: app/core/src/main/java/stirling/software/SPDF/controller/api/RotationController.java [hunk 5]
```diff
@@ -38,29 +38,43 @@ public class RotationController {
...
+
+    protected ResponseEntity<byte[]> respondPdf(PDDocument document, String filename)
+            throws IOException {
+        return WebResponseUtils.pdfDocToWebResponse(document, filename);
+    }
+
+    static void validateAngleMultipleOf90(int angle) {
+        if (angle % 90 != 0) {
+            throw ExceptionUtils.createIllegalArgumentException(
+                    "error.angleNotMultipleOf90", "Angle must be a multiple of 90");
```
Comment (balazs-szucs) [inline app/core/src/main/java/stirling/software/SPDF/controller/api/RotationController.java:77]: No reason to extract these methods. 

Author comment also, and lot of other places too.

File: app/core/src/main/java/stirling/software/SPDF/controller/api/RotationController.java [hunk 6]
```diff
@@ -38,29 +38,43 @@ public class RotationController {
...
             }
 
-            // Return the rotated PDF as a response
-            return WebResponseUtils.pdfDocToWebResponse(
-                    document,
-                    GeneralUtils.generateFilename(pdfFile.getOriginalFilename(), "_rotated.pdf"));
+            // Build response (via wrapper to allow stubbing)
+            String outName =
+                    GeneralUtils.generateFilename(pdfFile.getOriginalFilename(), "_rotated.pdf");
+            return respondPdf(document, outName);
```
Comment (balazs-szucs) [inline app/core/src/main/java/stirling/software/SPDF/controller/api/RotationController.java:60]: There is no changes in this file that is worth to keep around.

File: app/core/src/main/java/stirling/software/SPDF/controller/api/misc/MobileScannerController.java [hunk 7]
```diff
@@ -79,6 +79,14 @@ private ResponseEntity<Map<String, Object>> checkFeatureEnabled() {
...
     }
 
+    // Package-private on purpose: easy to unit-test without Spring context.
+    static String validateUploadFiles(List<MultipartFile> files) {
+        if (files == null || files.isEmpty()) {
+            return "No files provided";
+        }
+        return null; // null means OK
+    }
+
```
Comment (balazs-szucs) [inline app/core/src/main/java/stirling/software/SPDF/controller/api/misc/MobileScannerController.java:89]: Unnecesarily extracted method(?)

File: app/common/src/test/java/stirling/software/common/util/FileMonitorTest.java [hunk 8]
```diff
@@ -191,4 +191,111 @@ void testIsFileReadyForProcessing_DirectoryInsteadOfFile() throws IOException {
...
+                "path2KeyMapping",
+                new java.util.HashMap<Path, java.nio.file.WatchKey>());
+
+        // act
+        fileMonitor.trackFiles();
+
+        // assert mapping not empty (root registered)
+        var mapping = getPrivateField(fileMonitor, "path2KeyMapping");
+        assertTrue(
+                ((java.util.Map<?, ?>) mapping).size() > 0,
```
Comment (balazs-szucs) [inline app/common/src/test/java/stirling/software/common/util/FileMonitorTest.java:298]: FQN

File: app/common/src/test/java/stirling/software/common/util/FileMonitorTest.java [hunk 9]
```diff
@@ -191,4 +191,111 @@ void testIsFileReadyForProcessing_DirectoryInsteadOfFile() throws IOException {
...
+        fileMonitor.trackFiles();
+
+        // assert: newlyDiscoveredFiles contains the file
+        Set<Path> newly = getPrivateField(fileMonitor, "newlyDiscoveredFiles");
+        assertTrue(newly.contains(f), "existing file should be discovered on first run");
+
+        // assert: root registered
+        var mapping = getPrivateField(fileMonitor, "path2KeyMapping");
+        assertTrue(
+                ((java.util.Map<?, ?>) mapping).size() > 0, "root directory should be registered");
```
Comment (balazs-szucs) [inline app/common/src/test/java/stirling/software/common/util/FileMonitorTest.java:234]: FQN

File: app/common/src/test/java/stirling/software/common/util/propertyeditor/StringToArrayListPropertyEditorTest.java [hunk 10]
```diff
@@ -143,21 +147,80 @@ void testSetAsText_InvalidJson() {
...
+
         // Act
         editor.setAsText(json);
         Object value = editor.getValue();
 
         // Assert
-        assertNotNull(value, "Value should not be null");
-        assertInstanceOf(List.class, value, "Value should be a List");
+        assertNotNull(value);
+        assertInstanceOf(List.class, value);
```
Comment (balazs-szucs) [inline app/common/src/test/java/stirling/software/common/util/propertyeditor/StringToArrayListPropertyEditorTest.java:170]: I would instead:
```
RedactionArea area = list.get(0);
assertEquals(10.0, area.getX());
assertEquals("#FF0000", area.getColor());
```

File: .github/workflows/tauri-build.yml
Comment (balazs-szucs) [inline .github/workflows/tauri-build.yml:755]: Where did this come from? 😅 

Did you test this? What's the rationale?

File: .github/workflows/frontend-backend-licenses-update.yml
Comment (balazs-szucs) [inline .github/workflows/frontend-backend-licenses-update.yml:342]: So, I think you will need to bump gradle to version 9+, I don't think that gradle 8.X has Java 25 support

File: app/core/src/test/java/stirling/software/SPDF/exception/GlobalExceptionHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        assertEquals(MediaType.APPLICATION_PROBLEM_JSON, resp.getHeaders().getContentType());
+
+        ProblemDetail pd = resp.getBody();
+        assertNotNull(pd);
+
+        assertEquals("/api/method", pd.getProperties().get("path"));
+        assertNotNull(pd.getType());
+
+        assertEquals("TRACE", pd.getProperties().get("method"));
+        assertNotNull(pd.getProperties().get("supportedMethods"));
```
Comment (balazs-szucs) [inline app/core/src/test/java/stirling/software/SPDF/exception/GlobalExceptionHandlerTest.java:86]: I feel like with this and some of the other test in general there is lot of times you assert stuff like not null or does contain X etc... instead of asserting what is actually in the thing.

E.g., assertEquals(List.of("GET", "POST"), pd.getProperties().get("supportedMethods"));

This makes much sense than: assertNotNull(pd.getProperties().get("supportedMethods"));

This comes up pretty much quite a few times, but as a rule of thumb: the more specific the assertion can be, the better.

File: app/core/src/test/java/stirling/software/SPDF/exception/GlobalExceptionHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+        // handler-specific properties
+        assertEquals("q", pd.getProperties().get("parameterName"));
+        assertEquals("String", pd.getProperties().get("parameterType"));
+
+        // keep it simple: type exists
+        assertNotNull(pd.getType());
+
+        // title is set both as ProblemDetail.title and property "title"
+        assertEquals(pd.getTitle(), pd.getProperties().get("title"));
```
Comment (balazs-szucs) [inline app/core/src/test/java/stirling/software/SPDF/exception/GlobalExceptionHandlerTest.java:62]: Is asserting (potentially) non-RFC complaint behaviour I believe, though I'll do some reasearching. 

My intuition is that the ProblemDetail.title and property "title"  shouldn't be both set. It's duplication

File: app/core/src/test/java/stirling/software/SPDF/exception/GlobalExceptionHandlerTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ProblemDetail;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.HttpRequestMethodNotSupportedException;
+import org.springframework.web.bind.MissingServletRequestParameterException;
+
+import jakarta.servlet.http.HttpServletRequest;
+
+class GlobalExceptionHandlerTest {
```
Comment (balazs-szucs) [inline app/core/src/test/java/stirling/software/SPDF/exception/GlobalExceptionHandlerTest.java:20]: I feel like this quite light or? We obviously didn't have testing for this before the PR, but this can be great place to make more tests.

File: app/core/src/test/java/stirling/software/SPDF/controller/api/RotationControllerTest.java [hunk 14]
```diff
@@ -58,6 +62,62 @@ public void testRotatePDF() throws IOException {
         assertEquals(200, response.getStatusCode().value());
     }
 
+    @Test
+    public void testRotatePDFZeroAngle() throws IOException {
```
Comment (balazs-szucs) [inline app/core/src/test/java/stirling/software/SPDF/controller/api/RotationControllerTest.java:66]: is confusing.

Better test would be to: `testRotatePDF_ShouldMaintainExistingRotation_WhenAngleIsZero`

File: .github/workflows/build.yml
Comment (balazs-szucs) [inline .github/workflows/build.yml:348]: Do we need this?

File: .github/workflows/frontend-backend-licenses-update.yml
Comment (balazs-szucs) [inline .github/workflows/frontend-backend-licenses-update.yml:82]: if we already here maybe bump to node 24? That is the latest, right?
